### ISSUE TITLE
Issue with "Grouping operator ( )": (invalid  statements added)

### ISSUE THREAD

**Issue by stken2050** [mdn/content#9306]
MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Grouping

#### What information was incorrect, unhelpful, or incomplete?

>Notice in these examples that the left-to-right order of evaluation is still preserved. In other words, the order in which the operators evaluate has changed, but the order in which the operands evaluate has not. For example in this code:

`a() * (b() + c())
`
>The function a will be called before the function b, which will be called before the function c.

#### What did you expect to see?

Nothing, because the statement is invalid and confuses the readers especially for beginners. 

This statement has been added because of the argument:
The last line of the comment by @js-choi :
https://github.com/tc39/proposal-pipeline-operator/issues/227#issuecomment-926317660
>(Actually, perhaps I should also edit MDN’s documentation to make this clearer. Maybe this old parenthesized-expressions-are-not-always-evaluated-first rule is tripping up many people learning JavaScript.)

However, he has a confusion of concept.

This example is just to illustrate the [Eager evaluation strategy of JavaScript](https://en.wikipedia.org/wiki/Evaluation_strategy#Eager_evaluation) that has nothing to do with **Dependency graph defined by **Grouping operator ( )****

https://en.wikipedia.org/wiki/Dependency_graph
>In mathematics, computer science and digital electronics, a **dependency graph** is a directed graph representing dependencies of several objects towards each other. It is possible to derive an **evaluation order** or the absence of an **evaluation orde**r that respects the given dependencies from the dependency graph.

Remember "**evaluation order**" derived from the dependency graph or structure is completely different concept of "**time order**" of executions, and same goes to the false description in the page:

>Notice in these examples that the left-to-right **_order of evaluation_** is still preserved. In other words, **_the order in which the operators evaluate_** has changed, but **_the order in which the operands evaluate_** has not. For example in this code:

`a() * (b() + c())
`
>The function a will be called **_before_** the function b, which will be called **_before_** the function c.

#### Did you test this? If so, how?
 Yes.
I elaborated this:
[TC39/proposal-pipeline-operator Hack-style `|>` hijacks Grouping operator `( )`](https://dev.to/stken2050/the-current-tc39-proposal-pipeline-operator-hack-style-hijacks-grouping-operator-1dam)

Basically the pull request to add this invalid  statement is for the purpose of justifying the incompatibility of their own TC39 proposal. 

As invalid sentences, the description is harmful to every programmer including learner.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/javascript/reference/operators/grouping`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Grouping
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/operators/grouping/index.md
* Last commit: https://github.com/mdn/content/commit/b2f0a92a06f6dbf9c1a7c5de48152ca18875d168
* Document last modified: 2021-09-26T04:51:36.000Z

</details>

**Comment by stken2050** [comment]
>This was changed in #9243 by a TC39 delegate, and many TC39 delegates have weighed in and agree that the changed wording is an improvement. 

Then the process is now reconsidered because this statements include confusion of concept of Eager evaluation and evaluation order derived from Dependency graph.

The role of `()` in terms of both grouping operator and mathematics is to control dependency graph of the algebraic structure which has nothing to do with Eager evaluation in JavaScript.

>In JavaScript, "evaluation order" is NOT different from "time order" - they are synonyms. The sentence is valid.

I don't think so, and it's not you who judge this thing. Sorry.

**Comment by stken2050** [comment]
-------
#### What information was incorrect, unhelpful, or incomplete?

>Notice in these examples that the left-to-right order of evaluation is still preserved. In other words, the order in which the operators evaluate has changed, but the order in which the operands evaluate has not. For example in this code:

`a() * (b() + c())
`
>The function a will be called before the function b, which will be called before the function c.

------

The tricky word is: evaluation **"order"** ,  **"before"** or **"after"**

**In mathematics, when we use the term "order" "before" "after", does it mean time series? No.** 

What does matter is [dependency network](https://en.wikipedia.org/wiki/Dependency_network) structure.

Binary operation is merely a syntax-sugar of binary function, and when you chain them, you fundamentally compose the functions.

![image](https://lh5.googleusercontent.com/Ezak8AuE-a_wm92oraHwb5YNxTOXM0mMCaINgELAS6KNRAd9vSOFub1-hTrnWLC-tBKKutHTQOfzDX1KHQgrUURzjuUDVjvYqnN9SCW4uaJecGTDMydystqnrJGIeFml_z3sJNEE=s0)

For instance, when you enjoy + binary operation, you guys repeatedly told me "left to right", but fundamentally you do

![image](https://lh5.googleusercontent.com/ePh4ZCounDhhlLkx5sKhrOyeiAeeJHUTqDmKm7kAM_1Xvh03hm9hwYgKFAU37gjreDGqpogGayfmFWGVOUVM_kmGA5TOiLiDW-j5tg9ZIM9EZPniKSIcnDqwOae2jsiMwY-tPPWz=s0)

This is Fold

[https://en.wikipedia.org/wiki/Fold_(higher-order_function)](https://en.wikipedia.org/wiki/Fold_(higher-order_function))

There are both left and right side fold, and usually we use foldLeft and if it's monoid the result is the same of both side.

As you seen, this is a Graph. [Dependency graph](https://en.wikipedia.org/wiki/Dependency_graph)

>In mathematics, computer science and digital electronics, a dependency graph is a directed graph representing dependencies of several objects towards each other. It is possible to derive an **evaluation order** or the absence of an **evaluation order** that respects the given dependencies from the dependency graph.

**Remember "evaluation order" derived from the dependency graph or structure is completely different concept of "time order" of executions.**

**Comment by stken2050** [comment]
@teoli2003
@Rumyra

![image](https://user-images.githubusercontent.com/1316994/135003928-5a8e66d4-248d-4d78-97a6-08e6fe46fd31.png)

I'm afraid to mention that I don't think @ljharb is in the appropriate position to review this because he is the one of them who try to merge this invalid statement for the purpose of justifying the TC39 pipeline-operator Hack proposal. 

This statement has been added because of the argument:
The last line of the comment by @js-choi :
https://github.com/tc39/proposal-pipeline-operator/issues/227#issuecomment-926317660

> (Actually, perhaps I should also edit MDN’s documentation to make this clearer. Maybe this old parenthesized-expressions-are-not-always-evaluated-first rule is tripping up many people learning JavaScript.)

@ljharb has been the one who joined this discussion with very strong opinion, therefore he should be removed from the position to review here.

**Comment by Pokute** [comment]
It's clear that the reality of how JS behaves didn't match with the documentation.

Obviously it's infeasible to change how JS behaves without breaking a ton of existing code. Fixing documentation is simple and painless in comparison.

This is nothing new - the ECMAScript specification itself has been similarly changed to reflect the JS web reality multiple times IIRC.

**Comment by sideshowbarker** [comment]
Closing, after making an independent assessment that:

- The previous *“so that expressions with lower precedence can be evaluated before an expression with higher priority”* wording prior to the #9243 change was in fact problematic.
- The current *“so that operators with lower precedence can be evaluated before an operator with higher precedence”* wording after the #9243 change — and the new example and explanatory text added in that change (e.g., *“the order in which the *operators* evaluate has changed, but the order in which the *operands* evaluate has not”*) — are in fact a significant improvement.

**Comment by stken2050** [comment]
@Pokute I'm afraid to say you don't know what you are talking about, and simply ignored my explanation.

>It's clear that the reality of how JS behaves didn't match with the documentation.

The functionality of grouping operator that is to construct dependency graph
is independent from evaluation strategies of languages.

At the same time, the  order to resolve the dependency graph according to the grouping operator and binary-operator is independent from evaluation strategies of languages.

-------
Notice in these examples that the left-to-right order of evaluation is still preserved. In other words, the order in which the operators evaluate has changed, but the order in which the operands evaluate has not. For example in this code:

a() * (b() + c())

The function a will be called before the function b, which will be called before the function c.
----------

This gives a wrong impression as if eager evaluation of JavaScript spoils the principle of the functionality of grouping operator to construct dependency graph.

Well, it's ok someone feels the documentation is improved in the detail, however, as a whole, this document became more dangerous to have confusion of concepts of independence of eager evaluation of JavaScript and resolution of dependency graph constructed by grouping-operator `()` that is common in languages.

I must to repair this that should be not bad at all.

**Comment by sideshowbarker** [comment]
Locking the conversation, as this issue seems to be at risk of producing some unproductive back-and-forth without producing any new information that would merit re-opening the issue.

And to everybody involved in the conversation here: Let’s remind ourselves that discussions in this issue tracker a governed by the Mozilla Community Participation Guidelines, and let’s not in particular the following part of those guidelines:

https://www.mozilla.org/en-US/about/governance/policies/participation/#expected-behavior

> Value each other’s ideas, styles and viewpoints. We may not always agree, but disagreement is no excuse for poor manners. Be open to different possibilities and to being wrong. Be respectful in all interactions and communications, especially when debating the merits of different options.

Making a comment any part of which is at odds with those guidelines risks seeing that entire comment get deleted.

Finally, in general, when anyone wants to appeal some action taken on issues here in regard to the Mozilla Community Participation Guidelines, they can contact cpg-report@mozilla.com.
